overlord/ifacestate/apparmorprompting/common/common.go

```diff
@@ -336,18 +336,337 @@ func abstractPermissionsToAppArmorFilePermissions(iface string, permissions []st
 	return filePerms, nil
 }
 
-var allowablePathPatternRegexp = regexp.MustCompile(`^(/|(/[^/*{}]+)*(/\*|(/\*\*)?(/\*\.[^/*{}]+)?)?)$`)
+var (
+	// The following must be escaped as literals in a path pattern:
+	problematicChars = `\[\]`
+	// A single safe or escaped unsafe char in a path
+	safePathChar = fmt.Sprintf(`([^%s]|\\[%s])`, problematicChars, problematicChars)
+
+	// The following matches valid path patterns
+	allowablePathPattern       = fmt.Sprintf(`^/%s*$`, safePathChar)
+	allowablePathPatternRegexp = regexp.MustCompile(allowablePathPattern)
+)
+
+// Expands all groups in the given path pattern. Groups are enclosed by '{' '}'.
+// Returns a list of all the expanded path patterns, or an error if the given
+// pattern is invalid.
+func ExpandPathPattern(pattern string) ([]string, error) {
+	patternLen := len(pattern)
+	expanded := defaultPrefixes
```

Comment:
Yes, it must be a list with one item, as this is the list of previous patterns to which anything new should be appended. If it has length 0, there are no previous patterns, and thus the function would always return zero patterns. It's so that we can use the `expanded` variable throughout the function and continually add the previously-expanded components to it.

Essentially, whenever we encounter a group, we expand it, and then for each newly expanded path in that group, and for each previously expanded path in `expanded`, we adjoin the previous one to the new one. Then, at the end, we return `expanded`.

Do you think it would be clearer to rename `expanded` to `previouslyExpanded` or `expandedPrefixes`?

Reply:
I've renamed this to `currPrefixes`, and its behavior now matches that in `expandPathPatternFromIndex`.